File: src/algorithms/sets/generation/k-combinations/sources/k-combinations.py
# K-Combinations — Backtracking Generation
# Generates all C(n,k) subsets of exactly k elements from the input list.
# Time: O(k × C(n,k)) — generate C(n,k) combinations, each of length k
# Space: O(k × C(n,k)) — store all combinations


def k_combinations(elements: list[int], choose_k: int) -> list[list[int]]:
    result: list[list[int]] = []  # @step:initialize
    current_subset: list[int] = []  # @step:initialize

    def backtrack(start_idx: int) -> None:
        if len(current_subset) == choose_k:
            result.append(list(current_subset))  # @step:generate-subset
            return

        for elem_idx in range(start_idx, len(elements)):
            current_subset.append(elements[elem_idx])  # @step:initialize
            backtrack(elem_idx + 1)
            current_subset.pop()  # @step:backtrack

    backtrack(0)  # @step:initialize
    return result  # @step:complete

File: src/algorithms/sets/generation/k-combinations/sources/KCombinations.java
// K-Combinations — Backtracking Generation
// Generates all C(n,k) subsets of exactly k elements from the input array.
// Time: O(k × C(n,k)) — generate C(n,k) combinations, each of length k
// Space: O(k × C(n,k)) — store all combinations

import java.util.ArrayList;
import java.util.List;

public class KCombinations {

    public static List<List<Integer>> kCombinations(int[] elements, int chooseK) {
        List<List<Integer>> result = new ArrayList<>(); // @step:initialize
        List<Integer> currentSubset = new ArrayList<>(); // @step:initialize

        backtrack(elements, 0, chooseK, currentSubset, result); // @step:initialize
        return result; // @step:complete
    }

    private static void backtrack(
            int[] elements,
            int startIdx,
            int chooseK,
            List<Integer> currentSubset,
            List<List<Integer>> result) {
        if (currentSubset.size() == chooseK) {
            result.add(new ArrayList<>(currentSubset)); // @step:generate-subset
            return;
        }

        for (int elemIdx = startIdx; elemIdx < elements.length; elemIdx++) {
            currentSubset.add(elements[elemIdx]); // @step:initialize
            backtrack(elements, elemIdx + 1, chooseK, currentSubset, result);
            currentSubset.remove(currentSubset.size() - 1); // @step:backtrack
        }
    }
}
